### PR TITLE
toward types for vatWorker protocol

### PR DESCRIPTION
**Pull request by dckc**
This started as an experiment to figure out whether the compartmap used in xs-vat-worker has the same structure as the one used in endo. It does not. Reconciling the difference is still TODO.
https://github.com/Agoric/SES-shim/pull/458#issuecomment-694486578

I suppose `compartment-map-types.d.ts` should move to SES-shim, but I'm not sufficiently confident in my npm-fu to juggle work-in-progress in packages across monorepos.

Various bits should be squashed... maybe repackaged as different PRs...

### CODE REVIEW

**Review by michaelfig — APPROVED** [review]
Another step forward for typing!  I like it but would prefer not to lose the `asserts` type, which is legitimate typescript syntax that the JSDoc Typescript linter will accept sometime.

File: packages/SwingSet/src/capdata.js [hunk 1]
```diff
@@ -1,25 +1,32 @@
...
  * Assert function to ensure that something expected to be a capdata object
  * actually is.  A capdata object should have a .body property that's a string
  * and a .slots property that's an array of strings.
  *
- * @param {any} capdata  The object to be tested
+ * @param {unknown} capdata  The object to be tested
  * @throws {Error} if, upon inspection, the parameter does not satisfy the above
  *   criteria.
- * @returns {asserts capdata is CapData}
+ * @returns {void}
```
[the remark this answers is not in the document]
Comment (dckc) [reply packages/SwingSet/src/capdata.js]: I didn't realize this `@returns {asserts ...}` idiom was supported by `tsc`.

But now I see v3.7 adds [Assertion Functions](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions).

**Comment by dckc** [comment]
@michaelfig I can't figure out how to get this rest param to typecheck. I suspect some `tsc` version skew. What version of `tsc` do we use? I don't see it in `devDependencies` anywhere. Do we rely on git workflow to set the `tsc` version?

```
Error: src/vatWorker.js(10,27): error TS2370: A rest parameter must be of an array type.
```
https://github.com/Agoric/agoric-sdk/pull/1853/checks?check_run_id=1468224171#step:7:156

```js
/** @type { (first: string, ...args: unknown[]) => void } */
function workerLog(first, ...args) {
```

**Comment by michaelfig** [comment]
> @michaelfig I can't figure out how to get this rest param to typecheck. I suspect some `tsc` version skew. What version of `tsc` do we use? I don't see it in `devDependencies` anywhere. Do we rely on git workflow to set the `tsc` version?

I'm afk, but `tsc` is part of the `typescript` package.  I think that's specified at the top `package.json`.

**Comment by michaelfig** [comment]
> > @michaelfig I can't figure out how to get this rest param to typecheck. I suspect some `tsc` version skew. What version of `tsc` do we use? I don't see it in `devDependencies` anywhere. Do we rely on git workflow to set the `tsc` version?
> 
> I'm afk, but `tsc` is part of the `typescript` package. I think that's specified at the top `package.json`.

Guess not.  It should be.

**Comment by michaelfig** [comment]
> > @michaelfig I can't figure out how to get this rest param to typecheck. I suspect some `tsc` version skew. What version of `tsc` do we use? I don't see it in `devDependencies` anywhere. Do we rely on git workflow to set the `tsc` version?

> [It should be] specified at the top `package.json`.

IOW, please do a `(cd agoric-sdk && yarn add -W typescript)` and add the result to this PR.  If that's the last thing preventing CI from passing (never mind conventional-commits) I'll gladly review and we'll get it squared away.

**Comment by dckc** [comment]
> ... If that's the last thing preventing CI from passing (never mind conventional-commits) I'll gladly review and we'll get it squared away.

The commits are pretty messy. I'm inclined to squash some stuff.

**Comment by dckc** [comment]
@michaelfig fc8bfc5 fixes a build problem but introduces a bunch of problems with types. Ugh.

**Comment by warner** [comment]
@dckc says this is old and can be closed

**Pull request closed without merging**